flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EvictingWindowOperatorContractTest.java

```diff
@@ -0,0 +1,2654 @@
+		testNoTimerFiringForPurgedWindow(new EventTimeAdaptor());
+	}
+
+	@Test
+	public void testNoProcessingTimeFiringForPurgedWindow() throws Exception {
+		testNoTimerFiringForPurgedWindow(new ProcessingTimeAdaptor());
+	}
+
+	/**
+	 * Verify that we neither invoke the trigger nor the window function if a timer
+	 * for a non-existent window fires.
+	 */
+	private void testNoTimerFiringForPurgedWindow(final TimeDomainAdaptor timeAdaptor) throws Exception {
+
+		WindowAssigner<Integer, TimeWindow> mockAssigner = mockTimeWindowAssigner();
+		timeAdaptor.setIsEventTime(mockAssigner);
+		Trigger<Integer, TimeWindow> mockTrigger = mockTrigger();
+
+		@SuppressWarnings("unchecked")
+		InternalWindowFunction<Iterable<Integer>, List<Integer>, Integer, TimeWindow> mockWindowFunction =
+				mock(InternalWindowFunction.class);
+
+		KeyedOneInputStreamOperatorTestHarness<Integer, Integer, List<Integer>> testHarness =
+				createWindowOperator(mockAssigner, mockTrigger, 0L, intListDescriptor, mockWindowFunction);
+
+		testHarness.open();
+
+		timeAdaptor.advanceTime(testHarness, Long.MIN_VALUE);
+
+		when(mockAssigner.assignWindows(anyInt(), anyLong(), anyAssignerContext()))
+				.thenReturn(Arrays.asList(new TimeWindow(2, 4)));
+
+		assertEquals(0, testHarness.extractOutputStreamRecords().size());
+		assertEquals(0, testHarness.numKeyedStateEntries());
+
+		doAnswer(new Answer<TriggerResult>() {
+			@Override
+			public TriggerResult answer(InvocationOnMock invocation) throws Exception {
+				Trigger.TriggerContext context = (Trigger.TriggerContext) invocation.getArguments()[3];
+				// don't interfere with cleanup timers
+				timeAdaptor.registerTimer(context, 0L);
+				return TriggerResult.PURGE;
+			}
+		}).when(mockTrigger).onElement(Matchers.<Integer>anyObject(), anyLong(), anyTimeWindow(), anyTriggerContext());
+
+		testHarness.processElement(new StreamRecord<>(0, 0L));
+
+		assertEquals(0, testHarness.numKeyedStateEntries()); // not contents or state
+		assertEquals(2, timeAdaptor.numTimers(testHarness)); // timer and gc timer
+
+		timeAdaptor.advanceTime(testHarness, 0L);
+
+		// trigger is not called if there is no more window (timer is silently ignored)
+		timeAdaptor.verifyTriggerCallback(mockTrigger, never(), null, null);
+
+		verify(mockWindowFunction, never())
+				.apply(anyInt(), anyTimeWindow(), anyIntIterable(), EvictingWindowOperatorContractTest.<List<Integer>>anyCollector());
+
+		assertEquals(1, timeAdaptor.numTimers(testHarness)); // only gc timers left
+	}
+
+	@Test
+	public void testNoEventTimeFiringForPurgedMergingWindow() throws Exception {
+		testNoTimerFiringForPurgedMergingWindow(new EventTimeAdaptor());
+	}
+
+	@Test
+	public void testNoProcessingTimeFiringForPurgedMergingWindow() throws Exception {
+		testNoTimerFiringForPurgedMergingWindow(new ProcessingTimeAdaptor());
+	}
+
+
+	/**
+	 * Verify that we neither invoke the trigger nor the window function if a timer
+	 * for an empty merging window.
```

Comment:
seems like a missing word here: "if a timer _fires_ for an empty merging window"